### ISSUE TITLE
[Feature Request] More engine ?

### ISSUE THREAD

**Issue by AntoinePrv** [pytorch/ignite#215]
Hello,

Thank you for this library, I like the idea of handles. It makes the code more modular, separate the concepts and limit the size of these training loops functions!

I used ignite for the first time on custom data and loss and found the default [supervised trainers](https://github.com/pytorch/ignite/blob/c4563ddc67bee03fc616e87b84abd9b5e17e21cc/ignite/engine/__init__.py) to be very limited.

Of course implementing a new trainer is straightforward. My question is:
  - Do we want for ignite to offer a small collection of reasonable and general trainers?
  - Or do we not want ignite to go this extra step (just leave these ones for the sake of example) and leave it to the user?

In the first case, I have a couple ideas and would be happy to work on some PR.

**Comment by AntoinePrv** [comment]
The problem is related to #218, I worked on a problem where I wanted to give a mask (and another tensor) to the loss, on top of the usual `y`, `y_pred`.

So the `x, y = batch` from [_prepare_batch](https://github.com/pytorch/ignite/blob/c4563ddc67bee03fc616e87b84abd9b5e17e21cc/ignite/engine/__init__.py#L8) were a bit restrictive.

If I want to change just that function, I need to do something like:
```python
import ignite.engine
ignite.engine._prepare_batch = my_new_function
```
Which is ... not great to say the least.

I think having a bunch of functions `get_input(batch)`, `get_target(batch)` and `get_loss_kwargs(batch)` might help.
Of course we don't want to force the user to give such functions every time so we want to provide some default options. 
Maybe implementing the `supervised_trainer` as a class, the user could override only the part they want.

More generally, we could have two ways to create an engine: the `Engine(func)` that exists, and also by subclassing `Engine`. A special method would be the callback. This would give an API that feels like the pytorch `.forward()` in `torch.nn.Module`.


This one is a detail:
I think we could return the output of the model instead of the loss, that way, one can reuse it to compute other losses for cheap.

**Comment by vfdev-5** [comment]
@AntoinePrv could you detail please how did you used `supervised_trainer` with your `my_new_function`. I do not understand how the functions `get_input(batch)`, `get_target(batch)` and `get_loss_kwargs(batch)` can be helpful with a basic [`_update` function](https://github.com/pytorch/ignite/blob/master/ignite/engine/__init__.py#L29). I mean if batch is composed of `input`, `target` and `additional_data` how can it be then used in a general case in `_update` function ?

**Comment by AntoinePrv** [comment]
To make it simple, let's assume in my batch I have:
  - inputs sequences
 - a custom first hidden vector
 - target sequences
 - target mask
Represented as 4-tuple.

What I did was:
```python
import ignite.engine
from ignite._utils import convert_tensor

def _foo_prep_batch(batch, device=None):
    x = batch[:2]
    y = batch[2:]
    return convert_tensor(x, device=device), convert_tensor(y, device=device)

ignite.engine._prepare_batch =  _foo_prep_batch
```
And adapted my model and loss to deal with such input.

My idea was to do:
```python
class SupervidedTrainer:
    def __init__(model, optimizer, loss_fn, device=None):
        if device:
            model.to(device)

        self.model = model
        self.optimizer = optimizer
        self.loss_fn = loss_fn
        self.device = device


    def get_inputs(self, batch):
        return convert_tensor(batch[0], device=self.device)

    def get_targets(self, batch):
        return convert_tensor(batch[0], device=self.device)

    def get_loss_args(self, batch):
        return (), {}

    def __call__(self, engine, batch):
        model.train()
        optimizer.zero_grad()

        x = self.get_inputs(batch)
        y = self.get_inputs(batch)
        loss_args, loss_kwargs = self.get_loss_args(batch)

        y_pred = model(x)
        loss = loss_fn(y_pred, y, *loss_args, **loss_kwargs)
        loss.backward()
        optimizer.step()
        return y_pred
```
As the base trainer, that way if have extra args in my batch, I can just override ` get_loss_args` (same for inputs and targets).
Note: It's also possible to have the supervised trainer inherit from `Engine`.

**Comment by AntoinePrv** [comment]
**Nonetheless**, I just found out that Pytorch dataloaders support tuple of tuple of tensors, so I can just pack my inputs together and my options with my targets like:
```python
batch = ( (in_seq, hidden), (target, mask) )
```
Without having to modify `ignite.engine._prepare_batch`.
I can close the issue on that point if you don't want to add the extra complexity of this `SupervidedTrainer` class.

**Comment by vfdev-5** [comment]
@AntoinePrv thanks for the feedback! Glad that you could find the solution without too much modification :) 

> extra complexity of this `SupervidedTrainer` class.

Yes, looking at your example I think it brings more complexity than flexibility...

However, I was wondering whether it could be useful to add an argument to override `_prepare_batch` to the default `supervised_trainer`...

Yes, you can close the issue if you think that we are done with it and feel free to reopen if need
